From 80b680d4796d1ce913185dbaf4c36c6a64669e86 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Wed, 22 Feb 2017 11:03:04 +0100
Subject: [PATCH 1/2] Make cURL relocatable

This adds the ability to specify CA_BUNDLE paths that are relative to
the MSYS2 pseudo-root directory.

To make it truly relocatable, we use the path to the *cURL library*
instead of the path to the current .exe to determine the location of the
pseudo-root directory (allowing the .exe file to live completely outside
of the MSYS2 system, e.g. in $HOME/bin). This requires Win32 API
available in Windows XP & 2003 and later, well within the Windows
versions supported by Cygwin (and therefore MSYS2).

We also need to be extra careful to extend that path logic to the
ca-bundle.crt used to validate HTTPS *proxies*, not only HTTPS servers.

Note: This patch relies on the `pathtools.c` and `pathtools.h` files to
be copied from `mingw-w64-pathtools` into `lib/`.

Original-patch-by: Ray Donnelly <mingw.android@gmail.com>
Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 configure.ac         |  1 +
 lib/Makefile.inc     |  2 ++
 lib/curl_config.h.in |  3 +++
 lib/vtls/vtls.c      | 49 ++++++++++++++++++++++++++++++++++++++++++++
 4 files changed, 55 insertions(+)

diff --git a/configure.ac b/configure.ac
index cc16f4120..827d04948 100644
--- a/configure.ac
+++ b/configure.ac
@@ -4069,6 +4069,7 @@ dnl default includes
 ]
 )
 
+AC_DEFINE_UNQUOTED(CURL_BINDIR, "${prefix}/bin", [bindir])
 
 dnl Checks for typedefs, structures, and compiler characteristics.
 AC_C_CONST
diff --git a/lib/Makefile.inc b/lib/Makefile.inc
index 8b506febc..841e8a7ea 100644
--- a/lib/Makefile.inc
+++ b/lib/Makefile.inc
@@ -229,6 +229,7 @@ LIB_CFILES =         \
   noproxy.c          \
   openldap.c         \
   parsedate.c        \
+  pathtools.c        \
   pingpong.c         \
   pop3.c             \
   progress.c         \
@@ -361,6 +362,7 @@ LIB_HFILES =         \
   netrc.h            \
   noproxy.h          \
   parsedate.h        \
+  pathtools.h        \
   pingpong.h         \
   pop3.h             \
   progress.h         \
diff --git a/lib/curl_config.h.in b/lib/curl_config.h.in
index 668be5ea1..9c3139764 100644
--- a/lib/curl_config.h.in
+++ b/lib/curl_config.h.in
@@ -14,6 +14,9 @@
 /* If safe CA bundle search is enabled */
 #undef CURL_CA_SEARCH_SAFE
 
+/* Location of executable */
+#undef CURL_BINDIR
+
 /* Default SSL backend */
 #undef CURL_DEFAULT_SSL_BACKEND
 
diff --git a/lib/vtls/vtls.c b/lib/vtls/vtls.c
index df0449cbe..bb45551b0 100644
--- a/lib/vtls/vtls.c
+++ b/lib/vtls/vtls.c
@@ -81,6 +81,10 @@
 #include <Security/Security.h>
 #endif
 
+#if defined(__MINGW32__)
+#include "../pathtools.h"
+#endif
+
 /* The last #include files should be: */
 #include "../curl_memory.h"
 #include "../memdebug.h"
@@ -297,6 +301,15 @@ CURLcode Curl_ssl_easy_config_complete(struct Curl_easy *data)
 #if defined(CURL_CA_PATH) || defined(CURL_CA_BUNDLE)
   struct UserDefined *set = &data->set;
   CURLcode result;
+#ifdef __MINGW32__
+    const size_t path_max = PATH_MAX;
+#ifdef CURL_CA_PATH
+    char relocated_ca_path[path_max];
+#endif
+#ifdef CURL_CA_BUNDLE
+    char relocated_bundle[path_max];
+#endif
+#endif /* defined(__MINGW32__) */
 #endif
 
   if(Curl_ssl_backend() != CURLSSLBACKEND_SCHANNEL) {
@@ -305,16 +318,42 @@ CURLcode Curl_ssl_easy_config_complete(struct Curl_easy *data)
       sslc->native_ca_store = TRUE;
 #endif
 #ifdef CURL_CA_PATH
+#ifdef __MINGW32__
+    get_dll_path(relocated_ca_path, path_max);
+    strip_n_suffix_folders(relocated_ca_path, 1);
+    strncat(relocated_ca_path, "/", path_max - 1);
+    char *relative = get_relative_path(CURL_BINDIR, CURL_CA_PATH);
+    strncat(relocated_ca_path, relative, path_max - 1);
+    free((void*)relative);
+    simplify_path(relocated_ca_path);
+#endif /* defined(__MINGW32__) */
     if(!sslc->custom_capath && !set->str[STRING_SSL_CAPATH]) {
+#ifdef __MINGW32__
+      result = Curl_setstropt(&set->str[STRING_SSL_CAPATH], relocated_ca_path);
+#else
       result = Curl_setstropt(&set->str[STRING_SSL_CAPATH], CURL_CA_PATH);
+#endif /* defined(__MINGW32__) */
       if(result)
         return result;
     }
     sslc->primary.CApath = data->set.str[STRING_SSL_CAPATH];
 #endif
 #ifdef CURL_CA_BUNDLE
+#ifdef __MINGW32__
+    get_dll_path(relocated_bundle, path_max);
+    strip_n_suffix_folders(relocated_bundle, 1);
+    strncat(relocated_bundle, "/", path_max - 1);
+    char *relative = get_relative_path(CURL_BINDIR, CURL_CA_BUNDLE);
+    strncat(relocated_bundle, relative, path_max - 1);
+    free((void*)relative);
+    simplify_path(relocated_bundle);
+#endif /* defined(__MINGW32__) */
     if(!sslc->custom_cafile && !set->str[STRING_SSL_CAFILE]) {
+#ifdef __MINGW32__
+      result = Curl_setstropt(&set->str[STRING_SSL_CAFILE], relocated_bundle);
+#else
       result = Curl_setstropt(&set->str[STRING_SSL_CAFILE], CURL_CA_BUNDLE);
+#endif /* defined(__MINGW32__) */
       if(result)
         return result;
     }
@@ -353,8 +392,13 @@ CURLcode Curl_ssl_easy_config_complete(struct Curl_easy *data)
 #endif
 #ifdef CURL_CA_PATH
     if(!sslc->custom_capath && !set->str[STRING_SSL_CAPATH_PROXY]) {
+#ifdef __MINGW32__
+      result = Curl_setstropt(&set->str[STRING_SSL_CAPATH_PROXY],
+                              relocated_ca_path);
+#else
       result = Curl_setstropt(&set->str[STRING_SSL_CAPATH_PROXY],
                               CURL_CA_PATH);
+#endif /* defined(__MINGW32__) */
       if(result)
         return result;
     }
@@ -362,8 +406,13 @@ CURLcode Curl_ssl_easy_config_complete(struct Curl_easy *data)
 #endif
 #ifdef CURL_CA_BUNDLE
     if(!sslc->custom_cafile && !set->str[STRING_SSL_CAFILE_PROXY]) {
+#ifdef __MINGW32__
+      result = Curl_setstropt(&set->str[STRING_SSL_CAFILE_PROXY],
+                              relocated_bundle);
+#else
       result = Curl_setstropt(&set->str[STRING_SSL_CAFILE_PROXY],
                               CURL_CA_BUNDLE);
+#endif /* defined(__MINGW32__) */
       if(result)
         return result;
     }
